design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+# Title
+
+- Contribution Name: circuit-breaking
+- Implementation Owner: @Ashish-Bansal, Linkerd Team
+- Start Date: 2020-06-14
+- Target Date: N.A.
+- Linkerd Issue:
+  [linkerd/linkerd2#2846](https://github.com/linkerd/linkerd2/issues/2846)
+- Reviewers: @adleong
+
+## Summary
+
+[summary]: #summary
+
+We would like to implement circuit breaking functionality which would allow end
+users to configure mesh such that they can limit the impact of failures from
+faulty service over others and prevent cascading issues. By configuring circuit
+breaking, they will be able to remove failing service instances from load
+balancing and survive through it in a more graceful manner.
+
+## Problem Statement
+
+In a service mesh, services often make calls to other services. In case upstream
+service is unable to respond due to any reason, it's possible that these
+failures can cascade to other services in the cluster.
+
+If we are aware that specific service instance is not responding, we can call it
+unhealthy and it's better if we just stop sending "traffic"[0] to that instance
+rather than getting just another expected failure.
+
+---
+
+It would also be somewhat helpful in case of canary deployments when there's
+some regression and new build is raising failures.
+
+[Example](https://drive.google.com/open?id=1FxKi-egyhHumCvlHvTI4P114z7qOLPeN)
+
+---
+
+Users should also be able to set max-ejection-percentage which provides max
+number of service instances which can be removed from the load balancing for
+specific service.
+
+[Example](https://drive.google.com/open?id=1jSMaogdvzxjv0v2dBoL7OYGHmsCusXn-)
+
+---
+
+Network Parition - There can be lot of scenarios in network partition but to
+provide a glimpse how circuit breaking would behave in case of network issue -
+[Example](https://drive.google.com/open?id=154I2ZGlGc-rQJ0YVeek17dJy7raAvd-G)
+
+---
+
+Protocol Specific features -
+
+1. For HTTP/gRPC based communication, users should be able to configure circuit
+   breaking either per endpoint or whole service.
+   [Example](https://drive.google.com/open?id=1JK65u9Mnmkmy_X8fczeh6RO6O-vNiTtY)
+   User can configure ejection based on many different metrics. These metrics
+   would be calculated over a given time range.
+   - Success Rate
```

Comment:
Usecases - 

Success Rate - Most intuitive and generalised circuit breaking policy based on failure/success classification selected by user.

Latency based rate limiting - Let's say we have a service for which we cannot configure timeouts due to any reason. For example, few of the requests are expected to take longer than others. In this case, we cannot have circuit breaking on success rate. In these cases, we can rely on latency based circuit breaking.

Consecutive failures - In case, there are lot of failing requests due to some issue, we need to fail quickly instead of waiting for next time window evaluation.

Max requests - Hm, may be circuit breaking isn't an ideal place to add this. I'll remove this one.

---

I'll make above changes into the doc.